CSS: add units vw vh vmin vmax

ourairquality opened the pull request:
Discussed at https://github.com/buggins/coolreader/pull/246
 
Btw is it possible to disable the image scaling from the KOreader UI, to better test this code on KOreader?

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/koreader/crengine/413)
<!-- Reviewable:end -->

poire-z commented:
Thanks.

> Btw is it possible to disable the image scaling from the KOreader UI

Yes, KR frontend will do that soon, we're just in between PRs merged/not yet merged.
In the meantime, you can add into your (koreader-emulator...)/koreader/data/cr3.ini the values from:
https://github.com/koreader/koreader/pull/7204/files#diff-06ce311852251287dca96576aa4ecd91a4651152e3cf8f9f00ef5f71f1bf5382R203-R211

ourairquality commented:
Good suggestions thanks.

* Reordered the css types, but suggest the predicate function would be better but did not add that as percentage is sometimes excluded. The compiler will likely optimize to range tests. Did try to use parallel code style in the tests to make the refactoring opportunity more obvious.
* Updated those tests as suggested so that the new type is used in more paths.

poire-z reviewed (commented):
Don't forget to increment one of these when changes impact any of them:
https://github.com/koreader/crengine/blob/8ffbde4f84b568f517c843842ac0cce515ab1a5e/crengine/src/lvtinydom.cpp#L87-L91
The 2nd one in the case of this PR, as cached css_val_* numbers have been shifted/shuffled.

ourairquality commented:
Missed a few things there, thanks again. Hopefully addressed all of those.

Also added a predicate to test for the relative units, refactored, and cleaned up some of the associated logic - double negatives, empty conditional paths, conditional paths that had no effect.

poire-z commented:
> cleaned up some of the associated logic - double negatives, empty conditional paths, conditional paths that had no effect

Don't always assume it's best to do that. In some cases, I had been just following the specs logic/wording, and a refactor would lose the link with the specs (and empty conditional allows for putting verbose comments inside and so out of the main context, I sometimes prefer it that way).
Although for the cases you rewrote here, it's fine :) It was a bit convoluted, and I often had a hard time re-reading them (recently with the min/max-width/height) - and your rewrite is a LOT more readable :)

ourairquality commented:
Updated, thanks.

poire-z commented on crengine/src/lvrend.cpp:
```diff
@@ -2477,6 +2501,14 @@ int lengthToPx( css_length_t val, int base_px, int base_em, bool unspecified_as_
     return px;
 }
 
+bool is_length_relative_unit(css_length_t val)
+{
+    return (val.type == css_val_percent || val.type == css_val_em ||
+            val.type == css_val_ex || val.type == css_val_rem ||
+            val.type == css_val_vw || val.type == css_val_vh ||
+            val.type == css_val_vmin || val.type == css_val_vmax);
```
Really not better in include/cssdef.h ? Having that near the enum would allow going with `>` and `<` checks

ourairquality commented:
Looks like something broke with all the changes, let me look into it.

ourairquality commented:
Seems ok, sorry. Left is_length_relative_unit() as-is, it depends on css_length_t now. The predicate abstracts the multiple tests away for the developer, and the compiler can optimize to a range test. Using a numeric range tests on enums seems like over optimization, makes the code brittle, a foot gun. Anyway it's abstracted to a small function now, so you can change that easily in a follow up if that suits.

poire-z commented:
Looks good to me, no regression noticed on a few of my test files.

> Left is_length_relative_unit() as-is, it depends on css_length_t now. The predicate abstracts the multiple tests away for the developer, and the compiler can optimize to a range test

You haven't told me your thoughts about why or why not abstract it further away in include/cssdef.h ?

ourairquality commented:
Re ```is_length_relative_unit()``` it seems ok as-is, it is only used a few times in one file so that tips it towards just keeping it there for now, and it is next to ```lengthToPx()``` which it is often used with, but it does not seem worth wasting time overthinking that choice so just move it if you wish.

Could we explore the current choice of not applying the max-height?  Perhaps the utility of max-height is more obvious with the ```vh``` unit support. If a block height can be applied, then why not a max-height? I've appended a patch here to do this. Is there an example that you could point to that this breaks or that shows that this is not a great choice for CRengine? Could this as least be enabled for the viewport units? Here is a simple example to explore:

```
<div style="display: block; background: green; width: 20em; height: 20em; max-height: 60vh; max-width: 60vw"> </div>
```

ourairquality commented:
Updated the lengthToPx calculations for the viewport units to round to nearest. Just used division rather than shifting to make the rounding code pattern clear. Would you like the other unit calculations to round to nearest too in this PR?

poire-z commented:
> Could we explore the current choice of not applying the max-height? Perhaps the utility of max-height is more obvious with the vh unit support. If a block height can be applied, then why not a max-height? I've appended a patch here to do this. Is there an example that you could point to that this breaks or that shows that this is not a great choice for CRengine? 

I'm fine with this patch and limiting max-height support to just adjusting the value we got from CSS height/min-height to compute the `style_h` value that we will extend a block height if its content is lower. I just really didn't think of that.
(What I just don't want is ensuring max-height later, when the inner content height has been computed, and have it reduce the container height to be smaller than the inner content height, and have that inner content overflow its container on the Y-axis.)

Just a note: in [the specs](https://www.w3.org/TR/CSS2/visudet.html#min-max-heights), min-something is computed after max-something, so you might want to move your new block before the min-height one. (`style="height: 5; max-height: 2; min-height: 4"` should give 4, not 2).

And add a bit of the comment you removed, after the min/max-height computations:
`// Note: we'll always use the height needed to show this block content without overflowing`
or something like that to say that style_h is a min value.

> Just used division rather than shifting to make the rounding code pattern clear. Would you like the other unit calculations to round to nearest too in this PR?

I don't mind either way, it's just +/- 1 screen px that we hardly notice. But OK with that.
But you can just use `( ... + 128) >> 8` if you don't want to throw away our bitwise arithmetic shifts overthought micro-optimisations, but that could make it 2 operations.
I find ` /25600` a bit cryptic , `/256 /100` would be better). Try to decompose it so the original values stay visible and we understand it more easily, the compiler will probably optimize that. Or put the original human computation as a comment.

poire-z commented:
> > Would you like the other unit calculations to round to nearest too in this PR?

> I don't mind either way, it's just +/- 1 screen px that we hardly notice. But OK with that.

Yes, ok with that.
If you're adding half a unit before the division for vh/vm, you might as well do it for the others for consistency. `(...+128) / 256` looks fine, if the compiler is clever enough to notice that for /256 it can just do >>8.

No other comment about this PR, looks fine to me.

ourairquality commented:
Can I leave the rounding to another PR. I have a patch to propose but don't want to keep adding stuff on here, and there are other conflicting PRs, and rounding does not appear to be a priority matter. So perhaps this could be merge when convenient.
